docs/content/posts/2023-04-12-howto-migrate-to-datastream.md

```diff
@@ -0,0 +1,190 @@
+would have done on a DataSet unless it will not materialize the collection of data. The summing
+operation downstream is still done through
+a [ReduceFunction](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/functions/ReduceFunction.html)
+but this time the operator reduces the elements incrementally instead of receiving the rows as a
+Collection. So we store in the reduced row the partially aggregated sum. Due to incremental reduce,
+we also need to distinguish if we received an already reduced row (in that case, we read the
+partially aggregated sum) or a fresh row (in that case we just read the corresponding price field).
+
+### [Migrating the order by operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L172-L199)
+
+The sort of the datastream is done by applying
+a [KeyedProcessFunction](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/functions/KeyedProcessFunction.html)
+.
+
+But, as said above, the DataStream semantics are the ones of a streaming pipeline. The arriving data
+is thus considered infinite. As such we need to "divide" the data to have output times. For that we
+need to set a timer to output the resulting data. As we are in batch mode we work in processing
+time (and not in event time)
+and [set a timer to fire at Long.MAX_VALUE timestamp](https://github.com/echauchot/tpcds-benchmark-flink/blob/9c65e535bbd7f9c7f507e499c31c9280be2993ca/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L179)
+meaning that the timer will fire at the end of the batch.
+
+To sort the data, we store the incoming rows inside
+a [ListState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ListState.html)
+and sort them at output time, when the timer fires in
+the [onTimer()](https://javadoc.io/static/org.apache.flink/flink-streaming-java/1.16.0/org/apache/flink/streaming/api/functions/KeyedProcessFunction.html#onTimer-long-org.apache.flink.streaming.api.functions.KeyedProcessFunction.OnTimerContext-org.apache.flink.util.Collector-)
+callback.
+
+Another thing: to be able to use Flink state, we need to key the datastream beforeward even if there
+is no group by key because Flink state is designed per-key. Thus, we key by a fake static key so
+that there is a single state.
+
+### [Migrating the limit operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L201-L204)
+
+Like for the order by migration, here also we need to adapt to the streaming semantics in which we
+don't have the whole data at hand. Once again, we use the state API. We apply a mapper that stores
+the incoming rows into
+a [ValueState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ValueState.html)
```

Comment:
Why aren't you linking to the Flink javadocs at https://nightlies.apache.org/flink/flink-docs-release-1.16/api/java/?

Reply:
because at the time I wrote this article I did not know that javadocs were also in nightlies. Should be good for the other articles, I'll update all the links in this article.

Reply:
done